### ISSUE TITLE
_mm256_loadu_si256 only loads 128 bits when compiled with default cargo build --release

### ISSUE THREAD

**Issue by djsweet** [rust-lang/rust#52636]
The AVX2 intrinsic `_mm256_loadu_si256` fully loads all 256 bits from memory into the register when compiled without any optimization, but only loads 128 bits when compiled with the default `cargo build --release` option. This small program exhibits the issue:

```rust
use std::arch::x86_64;

fn main() {
    if is_x86_feature_detected!("avx2") {
        let load_bytes: [u8; 32] = [0x0f; 32];
        let lb_ptr = load_bytes.as_ptr();
        let reg_load = unsafe {
            x86_64::_mm256_loadu_si256(
                lb_ptr as *const x86_64::__m256i
            )
        };
        println!("{:?}", reg_load);
        let mut store_bytes: [u8; 32] = [0; 32];
        let sb_ptr = store_bytes.as_mut_ptr();
        unsafe {
            x86_64::_mm256_storeu_si256(sb_ptr as *mut x86_64::__m256i, reg_load);
        }
        assert_eq!(load_bytes, store_bytes);
    } else {
        println!("AVX2 is not supported on this machine/build.");
    }
}
```

When I run `cargo run`, this is the output:

```
   Compiling...
    Finished dev [unoptimized + debuginfo] target(s) in 0.33s
     Running `target/debug/avx2_bug_hunt`
__m256i(1085102592571150095, 1085102592571150095, 1085102592571150095, 1085102592571150095)
```

However, with `cargo run --release`, this is the output:
```
   Compiling...
    Finished release [optimized] target(s) in 0.26s
     Running `target/release/avx2_bug_hunt`
__m256i(1085102592571150095, 1085102592571150095, 0, 0)
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `[15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15]`,
 right: `[15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]`', src/main.rs:18:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

I am on macOS 10.13.6 with a Core i7 I7-4960HQ, and the output of `rustc --version --verbose` is

```
rustc 1.27.2
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-apple-darwin
release: 1.27.2
LLVM version: 6.0
```

Curiously, when inspecting the assembly of `main`, the call to `_mm256_loadu_si256` is not inlined, but instead generates this function:

```
avx2_bug_hunt`core::coresimd::x86::avx::_mm256_loadu_si256::hd7fc98ebefdce593:
avx2_bug_hunt[0x1000018a0] <+0>:  pushq  %rbp
avx2_bug_hunt[0x1000018a1] <+1>:  movq   %rsp, %rbp
avx2_bug_hunt[0x1000018a4] <+4>:  vmovaps %ymm0, (%rdi)
avx2_bug_hunt[0x1000018a8] <+8>:  popq   %rbp
avx2_bug_hunt[0x1000018a9] <+9>:  vzeroupper 
avx2_bug_hunt[0x1000018ac] <+12>: retq   
avx2_bug_hunt[0x1000018ad] <+13>: nopl   (%rax)
```

Note the `vzeroupper` instruction, which clears out the non-XMM registers. This is incorrect behavior, `_m256i` requires the full YMM register to be loaded unmodified. A similar spurious `vzeroupper` is also present in the assembly generated for `_mm256_storeu_si256`, but after the register is stored into memory.

**Comment by hellow554** [comment]
With `rustc` this only happens with `opt-level=3`, not 2.

**Comment by rkruppe** [comment]
> Note the vzeroupper instruction, which clears out the non-XMM registers. This is incorrect behavior, _m256i requires the full YMM register to be loaded unmodified. A similar spurious vzeroupper is also present in the assembly generated for _mm256_storeu_si256, but after the register is stored into memory.

I don't think this analysis is correct, the `loadu` function doesn't return the loaded value in a register (to avoid ABI mismatches) but rather returns it in memory (writing the contents of ymm to an address that's passed in, in this case the address of `reg_load`).  Indeed the assembly never even mentions ymm registers or AVX instructions outside of `loadu` and `storeu`, in `main` the 256 bit vectors are handled as a pair of xmm registers. It *can't* be using them, since the avx2 target feature is not enabled on the function. Incidentially, this target feature mismatch is also why `loadu` and `storeu` are not being inlined.

More importantly, this combined with #50154 is why *the store* doesn't work: `storeu` expects the value to be stored in a ymm register, but the caller doesn't have that register, so it passes the two halves of the vector in xmm0 and xmm1 instead. Normally rustc passes all vector arguments and return values through memory to avoid this problem, but as described in #50154 LLVM sometimes undoes this with disastrous consequences.

tl;dr duplicate of #50154

**Comment by djsweet** [comment]
@rkruppe You're right that this is related to #50154, but based on your comments here, we might have established the _true_ cause of these AVX bugs.

My use of `is_x86_feature_detected!("avx2")` is a bit of a red herring. I discovered this bug while attempting to use AVX2-specific intrinsics, but `_mm256_loadu_si256` is a part of AVX as per the [Intel Intrinsics Guide](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#!=undefined&expand=3166,3166&text=_mm256_loadu_si256), and is only enabled in `stdsimd` if the `avx` feature flag is enabled.

> the loadu function doesn't return the loaded value in a register (to avoid ABI mismatches) but rather returns it in memory (writing the contents of ymm0 to an address that's passed in, in this case the address of reg_load).

You appear to have `_mm256_loadu_si256` and `_mm256_storeu_si256` conflated. The only argument passed explicitly to `_mm256_loadu_si256`, `reg_load`, is used to _load_ the register, even as per the [`stdsimd` source](https://doc.rust-lang.org/src/core/up/stdsimd/coresimd/x86/avx.rs.html#1761). If the intent is to perform a memory-to-memory copy to avoid ABI mismatch, this has failed here, too. The assembly listing for `_mm256_loadu_si256` takes the input argument at `%rdi` and loads 256 bits from it into `%ymm0`. This aligns with the Rust-level definition of the function, and unless the `__m256` structs have special compiler support, would be what I would expect from reading that definition.

I find the idea that registers should be "demoted" to memory regions to fit an ABI circumspect. If the target ABI does not allow for the use of said registers, it would be more ergonomic for the compiler to fail early and often, notifying the user that their target ABI does not support the registers they are attempting to use.

**Comment by rkruppe** [comment]
> My use of is_x86_feature_detected!("avx2") is a bit of a red herring. I discovered this bug while attempting to use AVX2-specific intrinsics, but _mm256_loadu_si256 is a part of AVX as per the Intel Intrinsics Guide, and is only enabled in stdsimd if the avx feature flag is enabled.

I don't see how that's relevant to this issue.

> You appear to have _mm256_loadu_si256 and _mm256_storeu_si256 conflated. The only argument passed explicitly to _mm256_loadu_si256, reg_load, is used to load the register, even as per the stdsimd source. 

I am talking about the LLVM IR and assembly generated, not about the source code. However, it appears we're talking about different machine code: on the playground and in @djsweet's experiments, the body of `loadu` is essentially:

```
	vmovups	(%rsi), %ymm0
	vmovaps	%ymm0, (%rdi)
```

while your original post quotes a disassembly that contains only `vmovaps %ymm0, (%rdi)` ... huh, that's a store? That can't be right. Please double-check.

In any case, #50154 would also explain why `storeu` is miscompiled.

> unless the __m256 structs have special compiler support

They have, not just in the sense required for allocating them to ymm registers and using AVX instructions on them, but yes also in the sense you mean here: they are passed in memory rather than as immediates.

> I find the idea that registers should be "demoted" to memory regions to fit an ABI circumspect. If the target ABI does not allow for the use of said registers, it would be more ergonomic for the compiler to fail early and often, notifying the user that their target ABI does not support the registers they are attempting to use.

Unfortunately the ABI mismatch problems are real, difficult to solve, we can't very well ignore them, so this is the only feasible approach for the time being.

**Comment by djsweet** [comment]
@rkruppe The goofy AT&T ASM syntax always trips me up. In Intel-flavor mnemonics,

```
avx2_bug_hunt`core::coresimd::x86::avx::_mm256_load_si256::h86383795cdef4461:
avx2_bug_hunt[0x100000cc0] <+0>:  push   rbp
avx2_bug_hunt[0x100000cc1] <+1>:  mov    rbp, rsp
avx2_bug_hunt[0x100000cc4] <+4>:  vmovaps ymmword ptr [rdi], ymm0
avx2_bug_hunt[0x100000cc8] <+8>:  pop    rbp
avx2_bug_hunt[0x100000cc9] <+9>:  vzeroupper 
avx2_bug_hunt[0x100000ccc] <+12>: ret    
avx2_bug_hunt[0x100000ccd] <+13>: nop    dword ptr [rax]

avx2_bug_hunt`core::coresimd::x86::avx::_mm256_store_si256::h1b513ee5f14081f3:
avx2_bug_hunt[0x100000cd0] <+0>:  push   rbp
avx2_bug_hunt[0x100000cd1] <+1>:  mov    rbp, rsp
avx2_bug_hunt[0x100000cd4] <+4>:  vmovaps ymmword ptr [rdi], ymm0
avx2_bug_hunt[0x100000cd8] <+8>:  pop    rbp
avx2_bug_hunt[0x100000cd9] <+9>:  vzeroupper 
avx2_bug_hunt[0x100000cdc] <+12>: ret    
avx2_bug_hunt[0x100000cdd] <+13>: nop    dword ptr [rax]
```

For some reason, `load` and `store` are compiling identically on optimization level 3.

**Comment by alexcrichton** [comment]
I've confirmed that this is the same bug as https://github.com/rust-lang/rust/issues/50154, which is the same as upstream LLVM https://bugs.llvm.org/show_bug.cgi?id=37358 as @rkruppe mentioned

**Comment by alexcrichton** [comment]
@djsweet I don't personally know how we could fix this at the rustc level, but it may be good to ping the LLVM issue if you're interested in stirring up activity!
